### PR TITLE
fix(sql-udf): correctly handle `udf_binding_flag` & `udf_global_count`

### PR DESCRIPTION
**Pull request by xzhseh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

fixed a subtle but somewhat critical bug in the current sql udf implementation.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Just a note for docs to cover the named sql udf use case as demonstrated in the update test.

</details>

### CODE REVIEW

File: src/frontend/src/handler/create_sql_function.rs [hunk 1]
```diff
@@ -204,7 +208,7 @@ pub async fn handle_create_sql_function(
                     }
                 }
                 Err(e) => return Err(ErrorCode::InvalidInputSyntax(format!(
-                    "failed to conduct semantic check, please see if you are calling non-existence functions: {}",
+                    "failed to conduct semantic check, please see if you are calling non-existence functions or parameters\ndetailed error message: {}",
```
Comment (xzhseh) [inline src/frontend/src/handler/create_sql_function.rs:211]: This is just a current work around, I prefer a better hint display in the future, related: #14853.

File: src/frontend/src/binder/expr/function.rs [hunk 2]
```diff
@@ -241,10 +241,24 @@ impl Binder {
...
+
+                    // Only the top-most sql udf binding should unset the flag
+                    // Otherwise the subsequent binding may not be able to
+                    // find the corresponding context, consider the following example:
+                    // e.g., `select add_wrapper(a INT, b INT) returns int language sql as 'select add(a, b) + a';`
+                    // The inner `add` should not unset the flag, otherwise the `a` will be treated as
+                    // a normal column, which is then invalid in this context.
+                    if self.udf_context.global_count() == 0 {
+                        self.unset_udf_binding_flag();
+                    }
```
Comment (wangrunji0408) [inline src/frontend/src/binder/expr/function.rs:257]: Looks like `udf_binding_flag` can be inferred by `udf_global_counter != 0`. Can we remove `udf_binding_flag` now?

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

File: src/frontend/src/binder/expr/function.rs (hunk 2 above)
Comment (xzhseh) [reply src/frontend/src/binder/expr/function.rs:257]: I think so, let me do a tiny refactor.

**Pull request merged**